### PR TITLE
[config/configtls] Enable goleak check

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

This change enables goleak to check the config/configtls package for memory and goroutine leaks. There was originally a goroutine leak which is fixed by calling the newly added method `TLSServerSetting.Shutdown`.

The major change of this PR is to make the `*clientCAsFileReloader` object a member of the `TLSServerSetting` struct, as well as adding a public `Shutdown` method to `TLSServerSetting`. This was the cleanest way I could add a way to shut down the reloader, otherwise it will be a leaked goroutine. 

Also, I had to change `LoadTLSConfig()` to be a [pointer type receiver](https://gobyexample.com/methods), so that it could modify the given object by reference instead of only having the value copy of the `TLSServerSetting` object it was called on.

**Link to tracking Issue:** <Issue number if applicable>
#9165

**Testing:** <Describe what testing was performed and which tests were added.>
Goleak is passing successfully, as well as all existing tests.

### CODE REVIEW

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
Sorry about the go.mod conflict. Once that is fixed, this can be merged. Thank you!

**Review by jpkrohling — APPROVED** [review]

File: config/configtls/configtls.go
Comment (bogdandrutu) [inline config/configtls/configtls.go:345]: Strange logic that we create a reloader even when not starting, maybe split the logic? Separate PR would be preferred for that.

File: config/configtls/configtls.go
Comment (bogdandrutu) [inline config/configtls/configtls.go:333]: Since for the same `TLSServerSetting` we may call this multiple times, it may be better to actually return a `done` callback instead and have the reloader shutdown there.

Comment (jpkrohling) [reply config/configtls/configtls.go:333]: Great catch!

File: config/configtls/configtls.go
Comment (crobert-1) [reply config/configtls/configtls.go:345]: To make sure I understand, you're saying it's possible to get here even when the collector is not starting, so we should fix this logic to make sure the reloader only starts watching during start, am I following your thought?

File: config/configtls/configtls.go
Comment (crobert-1) [reply config/configtls/configtls.go:333]: So `done` would be a function that shuts down the reloader, and is called by the user on shutdown or when they no longer need the current TLS config, is that correct?

File: config/configtls/clientcasfilereloader.go
Comment (crobert-1) [inline config/configtls/clientcasfilereloader.go:135]: I don't think we should return an error if the shutdown channel is nil.

[From documentation:](https://pkg.go.dev/go.opentelemetry.io/collector/component#Component)
```
	// This method must be safe to call:
	//   - without Start() having been called
	//   - if the component is in a shutdown state already
```
Maybe I'm misinterpreting what `safe to call` means, but I don't think we should return an error unless there's an issue shutting down. I don't think we should error if shutdown is successful. (Which I believe is the second point from docs here).

Even though this itself isn't specifically a collector component, I believe docs still apply as it's a part of collector shutdown.

File: .chloggen/goleak_configtls.yaml
Comment (crobert-1) [inline .chloggen/goleak_configtls.yaml:19]: Feedback is welcome here, I want to make sure the action needed by users is as clear as possible.

File: config/configtls/configtls.go
Comment (crobert-1) [reply config/configtls/configtls.go]: I believe I've implemented what was requested here 👍

Comment (crobert-1) [reply config/configtls/configtls.go]: Looking more at consumers of `TLSServerSetting.LoadTLSConfig`, it looks like there isn't a simple way to keep track of the lifecycle of these objects in a uniform way externally. It may make sense to keep track internally and then call all of the shutdown methods on `TLSServerSetting.Shutdown`. This will make it so this is no longer a breaking change, and the only extra requirement on users is calling the shutdown method here.

File: .chloggen/goleak_configtls.yaml
Comment (crobert-1) [reply .chloggen/goleak_configtls.yaml]: My new suggested resolution is a `bug_fix` instead of `breaking`, so this is no longer relevant.

File: config/configtls/configtls.go
Comment (crobert-1) [reply config/configtls/configtls.go:335]: I implemented my proposed solution and added tests.

**Comment by crobert-1** [comment]
@jpkrohling and @bogdandrutu: I believe I've addressed the concerns mentioned, another look would be appreciated when you're able!

File: config/configtls/clientcasfilereloader.go
Comment (TylerHelmuth) [reply config/configtls/clientcasfilereloader.go:135]: I think your change makes sense.  I interpret that comment to mean "calling shutdown on something not started should not error".

File: config/configtls/configtls.go
Comment (TylerHelmuth) [inline config/configtls/configtls.go:381]: Goland yells at me when I mix receiver types. I believe best practice is not to mix types, so if this needs to be a pointer I think all the methods need updating.

File: config/configtls/configtls.go
Comment (TylerHelmuth) [inline config/configtls/configtls.go:412]: Since this value is public, I believe it is possible for it to be set to true when calling `LoadTLSConfig`, which will cause some `shutdowns` funcs to be remembered, then for `ReloadClientCAFile` to be set to false, and then `Shutdown` to be called resulting in the `shutdown` funcs being executed.

I think it would be best to always try to loop through the list and invoke anything inside.

File: config/configtls/configtls.go
Comment (crobert-1) [reply config/configtls/configtls.go:381]: Done 👍

File: config/configtls/configtls.go
Comment (crobert-1) [reply config/configtls/configtls.go:412]: Good catch, thanks! I've updated.

File: receiver/otlpreceiver/otlp.go
Comment (TylerHelmuth) [inline receiver/otlpreceiver/otlp.go:201]: I'd prefer `errors.Join` instead of `multierr`.

Comment (crobert-1) [reply receiver/otlpreceiver/otlp.go:201]: Agreed, updated.

Comment (TylerHelmuth) [reply receiver/otlpreceiver/otlp.go:201]: If you don't want the resulting error to be nested I believe you need to keep a slice of errors and do the `errors.Join` once at the end.  I'm not sure here if that matters or not.

Comment (crobert-1) [reply receiver/otlpreceiver/otlp.go:201]: Thanks for catching that, I've updated.

File: receiver/otlpreceiver/otlp.go
Comment (TylerHelmuth) [inline receiver/otlpreceiver/otlp.go:198]: Looking at this some more I really don't like that any component that uses `configtls` has to add this check.  I am worried about components not realizing that the `configtls.TLSSetting` embedded in `confighttp.ServerConfig` needs a `Shutdown` call.

I think it is out of scope of this PR, which is only trying to solve a leak added by existing functionality`, but I think we should take another look at how `config` modules should handle goroutines and `Shutdown`.

**Review by TylerHelmuth — APPROVED** [review]

**Comment by yurishkuro** [comment]
Q: why do we implement reloading of client CA file with a background goroutine instead of using the same approach as `certReloader` which does it on-demand when GetCertificate is called?

https://github.com/open-telemetry/opentelemetry-collector/blob/ede9e304314dddbdeb10b5ee06b1a49ab88e4f69/config/configtls/configtls.go#L140

This would avoid a goroutine leak and the unpleasant side-effects of having to call Shutdown on a "config" object.

**Comment by mx-psi** [comment]
@crobert-1 What's the status of this PR? Does it need more reviews? Do you need to fix something? (No rush, just trying to clean up our backlog)

**Comment by crobert-1** [comment]
> @crobert-1 What's the status of this PR? Does it need more reviews? Do you need to fix something? (No rush, just trying to clean up our backlog)

Just waiting for reviews 👍

**Comment by mx-psi** [comment]
Can you take a look at Yuri's question above?

**Comment by mx-psi** [comment]
@crobert-1 Mind taking a look at https://github.com/open-telemetry/opentelemetry-collector/pull/9220#issuecomment-2119380494 ?

**Comment by crobert-1** [comment]
> @crobert-1 Mind taking a look at [#9220 (comment)](https://github.com/open-telemetry/opentelemetry-collector/pull/9220#issuecomment-2119380494) ?

My apologies for the delay, thanks for the ping! I'm taking a look at this. There's no context from the [original issue](https://github.com/open-telemetry/opentelemetry-collector/issues/6524) or [implementing PR](https://github.com/open-telemetry/opentelemetry-collector/pull/6708) why this design was chosen. However, the issue says it should be done like the `certReloader`, which, as @yurishkuro pointed out, is done in a lazy manner on `Get`.

I'm working on implementing this to simply read the file on `Get` rather than the proactive watching approach that's currently implemented. If it works properly, I'll open a new PR instead of this approach.

_Marking as draft until I get a chance to properly prove the alternative approach will work._

**Comment by akorompai** [comment]
> > @crobert-1 Mind taking a look at [#9220 (comment)](https://github.com/open-telemetry/opentelemetry-collector/pull/9220#issuecomment-2119380494) ?
> 
> My apologies for the delay, thanks for the ping! I'm taking a look at this. There's no context from the [original issue](https://github.com/open-telemetry/opentelemetry-collector/issues/6524) or [implementing PR](https://github.com/open-telemetry/opentelemetry-collector/pull/6708) why this design was chosen. However, the issue says it should be done like the `certReloader`, which, as @yurishkuro pointed out, is done in a lazy manner on `Get`.
> 
> I'm working on implementing this to simply read the file on `Get` rather than the proactive watching approach that's currently implemented. If it works properly, I'll open a new PR instead of this approach.
> 
> _Marking as draft until I get a chance to properly prove the alternative approach will work._

@crobert-1 Sorry for all the troubles my PR is causing, I didn't realize that the shutdown concern wasn't addressed before it got merged  (https://github.com/open-telemetry/opentelemetry-collector/pull/6708#discussion_r1088291202)
Please note that a dedicated client `certReloader` was implemented in a previous PR, but got rejected in favor of a fs watcher solution (https://github.com/open-telemetry/opentelemetry-collector/pull/6525#issuecomment-1322557224)

**Pull request closed without merging**
